### PR TITLE
[JUJU-2749] upgrade a resource without a new charm revision available.

### PR DESCRIPTION
**Pull request by hmlanigan**
Allow for upgrading a resource even if the charm used by an application will not be upgraded. Three small changes were needed to accomplish this. 

The refresh (update-charm) cli logic on failures related to ErrAlreadyUpToDate had to change. It's not an immediate failure. If there are resources to update or a charm's channel to change, still call SetCharm. Let the user know what is happening.

Update the logic around determination of whether a resource requires an upgrade by using the resource data available in the repository. When upgrading a resource required a charm revision change, this wasn't as important. However with the first change, ensure that we only pass along resources that must be updated. The logic in this part is complex.

SetCharm must to allow for pending resource resolution independent of a charm url change. Including incrementing CharmModifiedVersion if it wasn't done earlier in the method. CharmModifiedVersion is what triggers the uniter to call the upgrade charm hook if the charm url didn't change.

Finally added more integration tests to cover:

1. Refresh a resource without a new charm revision.
2. Refresh a resource, specifying a different revision than deployed, without a new charm revision.
3. Deploy with an uploaded resource, refresh fails to refresh the resource as it's been uploaded.

## QA steps

Depending on which version of ubuntu your machine is running, you may need to run `juju model-default default-series=focal`. Some of these tests use Xenial by default. 

Local charm scenarios:

1. Deploy a local charm without resources. Edit the metadata.yaml file to add a resource. Refresh the charm without uploading a resource file. It should fail.
2. Refresh the local charm from scenario 1 again, this time supplying the resource on the cli: --resource <resource-name>=<path-to-file>. It should succeed.

Use `juju resources <application-name> --format json | jq` to validate resources

```sh
(cd tests ; ./main.sh -v resources)
```

Test with k8s charms and oct-images as well, my attempt was here: https://bugs.launchpad.net/juju/+bug/1993319/comments/3

## Bug reference

Along the road to fixing: https://bugs.launchpad.net/juju/+bug/1993319 . Resources couldn't be updated without a charm revision change before this at all. The next step will be to ensure the bug related to upgrading oci-images is fixed.

### CODE REVIEW

File: cmd/juju/application/utils/utils.go [hunk 1]
```diff
@@ -98,15 +105,43 @@ func getResources(
...
 
-func filterResources(
+// getAvailableRepositoryResources gets the current resources for this
+// charm available in the repository.
+func getAvailableRepositoryResources(newCharmID application.CharmID, repositoryResourceLister CharmClient) (map[string]charmresource.Resource, error) {
+	if repositoryResourceLister == nil {
+		// not required for local charms
+		return nil, nil
+	}
+	available, err := repositoryResourceLister.ListCharmResources(newCharmID.URL, newCharmID.Origin)
```
Comment (wallyworld) [inline cmd/juju/application/utils/utils.go:115]: Does this account for published resources only? ie does it ignore uploaded resources that have not yet been published?

File: cmd/juju/application/utils/utils.go [hunk 2]
```diff
@@ -118,38 +153,87 @@ func filterResources(
...
-	cur, ok := current[res.Name]
-	if !ok {
+	if !curFound {
 		// If there's no information on the server, there might be a new resource added to the charm.
-		if newCharmURL.Schema == charm.Local.String() {
-			return false, errors.NewNotValid(nil, fmt.Sprintf("new resource %q was missing, please provide it via --resource", res.Name))
-		}
-
-		logger.Tracef("resource %q does not exist in controller, so it will be uploaded", res.Name)
+		logger.Tracef("resource %q does not exist in controller, so it will be uploaded", resName)
```
Comment (wallyworld) [inline cmd/juju/application/utils/utils.go:190]: s/uploaded/upgraded?

File: cmd/juju/application/utils/utils.go [hunk 3]
```diff
@@ -118,38 +153,87 @@ func filterResources(
...
-	// upgrade-charm.
-	if _, ok := uploads[res.Name]; ok {
-		logger.Tracef("%q provided to upgrade existing resource", res.Name)
+// resource.
+//
+// Upgrade when:
+//  1. Resources specified as a local file on the cli to be uploaded.
+//  2. Resources specified as resource revision on the cli, which are
+//     not currently in use.
+//  3. Upstream has a new resource available than what is currently used.
```
Comment (wallyworld) [inline cmd/juju/application/utils/utils.go:162]: s/new/newer

File: cmd/juju/application/utils/utils.go [hunk 4]
```diff
@@ -118,38 +153,87 @@ func filterResources(
...
+//  4. A new upstream resource is available.
```
Comment (wallyworld) [inline cmd/juju/application/utils/utils.go:163]: so long as it has been published right?

File: cmd/juju/application/utils/utils.go (hunk 1 above)
Comment (hmlanigan) [reply cmd/juju/application/utils/utils.go:115]: This is what the repository (charmhub) says are current resources revisions for the given channel. The resourceLister queries state to see what the current resources are. Uploaded resources are verified in shouldUpgradeResource.

File: cmd/juju/application/utils/utils.go (hunk 4 above)
Comment (hmlanigan) [reply cmd/juju/application/utils/utils.go:163]: yes, the repositoryResourceLister won't be given the data if it hasn't been published.

File: cmd/juju/application/utils/utils.go (hunk 2 above)
Comment (hmlanigan) [reply cmd/juju/application/utils/utils.go:190]: The charm has introduced a new resource that the application hasn't used before.  The user hasn't listed a file on the cli to use for it which would be uploaded. Maybe downloaded or keep the same?

File: tests/suites/resources/refresh.sh [hunk 5]
```diff
@@ -0,0 +1,124 @@
...
+	juju refresh juju-qa-test
+
+	juju config juju-qa-test foo-file=true
+	wait_for "resource line one: testing two." "$(workload_status juju-qa-test 0).message"
+  juju resources juju-qa-test --format json | jq -S '.resources[0] | .[ "revision"] == "2"'
+
+	destroy_model "test-${name}"
+}
+
+run_resource_refresh_no_new_charm_rev_supply_res_rev() {
```
Comment (hmlanigan) [inline tests/suites/resources/refresh.sh:52]: New test

File: tests/suites/resources/refresh.sh [hunk 6]
```diff
@@ -0,0 +1,124 @@
+run_resource_refresh() {
```
Comment (hmlanigan) [inline tests/suites/resources/refresh.sh:1]: existing test from upgrade.sh

File: tests/suites/resources/refresh.sh [hunk 7]
```diff
@@ -0,0 +1,124 @@
...
+	juju refresh juju-qa-test --resource foo-file=3
+
+	juju config juju-qa-test foo-file=true
+	wait_for "resource line one: testing three." "$(workload_status juju-qa-test 0).message"
+  juju resources juju-qa-test --format json | jq -S '.resources[0] | .[ "revision"] == "3"'
+
+	destroy_model "test-${name}"
+}
+
+run_resource_no_upgrade_after_upload() {
```
Comment (hmlanigan) [inline tests/suites/resources/refresh.sh:80]: new test

File: tests/suites/resources/refresh.sh [hunk 8]
```diff
@@ -0,0 +1,124 @@
...
+	juju refresh juju-qa-test --channel 2.0/stable
+	wait_for "juju-qa-test" "$(charm_channel "juju-qa-test" "2.0/stable")"
+
+	juju config juju-qa-test foo-file=true
+	wait_for "resource line one: testing one." "$(workload_status juju-qa-test 0).message"
+
+	destroy_model "test-${name}"
+}
+
+run_resource_refresh_no_new_charm_rev() {
```
Comment (hmlanigan) [inline tests/suites/resources/refresh.sh:26]: new test

File: cmd/juju/application/utils/utils.go (hunk 2 above)
Comment (wallyworld) [reply cmd/juju/application/utils/utils.go:193]: I think just tweak the message to avoid any ambiguity

File: state/application.go [hunk 9]
```diff
@@ -84,9 +84,11 @@ type applicationDoc struct {
 	// be relatively straight forward, but very time consuming.
 	// When moving to CharmHub or removing CharmStore from Juju it should be
 	// tackled then.
-	CharmURL             *string      `bson:"charmurl"`
-	Channel              string       `bson:"cs-channel"`
-	CharmOrigin          *CharmOrigin `bson:"charm-origin"`
+	CharmURL    *string      `bson:"charmurl"`
+	Channel     string       `bson:"cs-channel"`
```
Comment (wallyworld) [inline state/application.go:88]: sad panda that we can't easily get rid of "cs"

**Review by wallyworld — APPROVED** [review]
Thank you. I think this looks good to go.

File: state/application.go (hunk 9 above)
Comment (hmlanigan) [reply state/application.go:88]: It's gone from here in 3.1, this is 2.9 where it's still needed. This is a spacing change with the comment.

**Pull request merged**